docs/src/adr/023adr-trace-evaluation.md

```diff
@@ -0,0 +1,75 @@
+# ADR-023: Trace evaluation
+
+| authors                                | last revised    |
+| -------------------------------------- | --------------: |
+| Jure Kukovec                           | 2022-09-28      |
+
+**Table of Contents**
+
+- [Summary (Overview)](#summary)
+- [Context (Problem)](#context)
+- [Options (Alternatives)](#options)
+- [Solution (Decision)](#solution)
+- [Consequences (Retrospection)](#consequences)
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+In the context of improving usability\
+facing difficulties understanding counterexample traces\
+we decided to implement trace evaluation\
+to achieve a better user experience\
+accepting the development costs.\
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+As explained in [#1845](https://github.com/informalsystems/apalache/issues/1845), one often runs into the problem of unreadable counterexamples; 
+for complex specifications, it is often the case that either the sate contains many variables, or the invariant contains many conjunctive components (or both).
+In that case, trying to determine exactly why e.g. `A1 /\ ... /\ An` was violated basically boils down to manually evaluating each formula `Ai` with whatever variables are present in the state.
+This is laborious and error-prone, and should be automated.
+
+
+## Options
+
+<!-- Communicate the options considered.
+     This records evidence of our circumspection and documents the various alternatives
+     considered but not adopted.
+-->
+1. Call REPL in each state:
+    - No convenient REPL integration at the moment
+    - No clear way of saving outputs
+1. Encode trace traversal as an Apalache problem and use the solver
+    - No additional rules or IO needed
+
+## Solution
+
+<!-- Communicates what solution was decided, and it is expected to solve the
+     problem. -->
+
+We choose option (2). 
+
+Suppose we are given a trace `s0 -> s1 -> ... -> sn` over variables `x1,...,xk` as well as expressions `E1,...,Em`, such that all free variables of `E1,...,Em` are among `x1,...,xk`. W.l.o.g. assume all constants are inlined.
+
+The above defines a trace `t0 -> t1 -> ... -> tn` over variables `v1,...,vm`, such that 
+```
+ti.vj = Ej[si.x1/x1,...,si.xk/xk]
+```
```

Comment:
Could you use underscores in the variable names, e.g., `t_i`, `v_j`. It's hard to parse the current write up.

Reply:
I can, but I don't think markdown prettyprints those.

Reply:
I think you can use latex, in case that would help: https://github.blog/2022-05-19-math-support-in-markdown/

Reply:
> I think you can use latex, in case that would help: https://github.blog/2022-05-19-math-support-in-markdown/

Yeah, that's a very cool feature. But the ADRs are going via mdbook, so this would not work I suppose? We had some other special support for math environment.

Reply:
I used some mdbook-latex library in https://apalache.informal.systems/docs/tutorials/symbmc.html.
In hindsight, probably a massive waste of time, since I reckon like 3 people ever read that.